File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/DownloadPolicy.kt
package ai.quenderin.core

enum class NetworkStatus { WIFI, CELLULAR, NONE }

/**
 * Whether a large model download may proceed on the current connection. A multi-GB pull
 * on cellular can cost real money and stall — default to Wi-Fi-only, never start with no
 * connection. Pure decision layer (tested); the live status comes from the app via
 * `ConnectivityManager`. Mirrors iOS `DownloadPolicy`.
 */
enum class DownloadPolicy {
    WIFI_ONLY,
    WIFI_OR_CELLULAR;

    fun allows(status: NetworkStatus): Boolean = when {
        status == NetworkStatus.NONE -> false
        this == WIFI_ONLY && status == NetworkStatus.CELLULAR -> false
        else -> true
    }

    /** Why a download is being held back, in the user's words — or null if allowed. */
    fun reason(status: NetworkStatus): String? {
        if (allows(status)) return null
        return when (status) {
            NetworkStatus.NONE ->
                "No internet connection. Connect to Wi-Fi to download your model before going offline."
            NetworkStatus.CELLULAR ->
                "You're on cellular. This model is large — connect to Wi-Fi, or allow cellular downloads in settings."
            NetworkStatus.WIFI -> null
        }
    }
}

File: apple/QuenderinKit/Sources/QuenderinKit/DownloadPolicy.swift
import Foundation

public enum NetworkStatus: Sendable, Equatable {
    case wifi
    case cellular
    case none
}

/// Whether a large model download may proceed on the current connection.
/// A multi-GB pull on cellular can cost real money and stall — default to
/// Wi-Fi-only, and never start with no connection.
public enum DownloadPolicy: Sendable, Equatable {
    case wifiOnly
    case wifiOrCellular

    public func allows(_ status: NetworkStatus) -> Bool {
        switch (self, status) {
        case (_, .none): return false
        case (.wifiOnly, .cellular): return false
        default: return true
        }
    }

    /// Why a download is being held back, in the user's words — or nil if allowed.
    public func reason(for status: NetworkStatus) -> String? {
        guard !allows(status) else { return nil }
        switch status {
        case .none:
            return "No internet connection. Connect to Wi-Fi to download your model before going offline."
        case .cellular:
            return "You're on cellular. This model is large — connect to Wi-Fi, or allow cellular downloads in settings."
        case .wifi:
            return nil
        }
    }
}

#if canImport(Network)
import Network

/// Live network status via `NWPathMonitor`. Device-coupled (real connectivity),
/// so it isn't unit-tested; `DownloadPolicy` is the tested decision layer.
public final class LiveNetworkMonitor: @unchecked Sendable {
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "ai.quenderin.network")
    private let lock = NSLock()
    private var _status: NetworkStatus = .none

    public var status: NetworkStatus {
        lock.lock(); defer { lock.unlock() }
        return _status
    }

    public init() {
        monitor.pathUpdateHandler = { [weak self] path in
            let status: NetworkStatus
            if path.status == .satisfied {
                status = path.usesInterfaceType(.cellular) ? .cellular : .wifi
            } else {
                status = .none
            }
            self?.lock.lock(); self?._status = status; self?.lock.unlock()
        }
        monitor.start(queue: queue)
    }

    deinit { monitor.cancel() }
}
#endif
